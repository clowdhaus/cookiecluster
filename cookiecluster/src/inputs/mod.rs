pub(crate) mod add_on;
pub(crate) mod ami;
pub(crate) mod compute;
pub(crate) mod instance;
pub(crate) mod version;

pub use add_on::AddOn;
use anyhow::Result;
use dialoguer::{Confirm, Input, MultiSelect, Select, theme::ColorfulTheme};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Inputs {
  pub(crate) accelerator: compute::AcceleratorType,
  pub(crate) add_ons: Vec<add_on::AddOn>,
  pub(crate) ami_type: ami::AmiType,
  pub(crate) cluster_endpoint_public_access: bool,
  pub(crate) cluster_name: String,
  pub(crate) cluster_version: version::ClusterVersion,
  pub(crate) control_plane_subnet_filter: String,
  pub(crate) compute_scaling: compute::ScalingType,
  pub(crate) cpu_arch: compute::CpuArch,
  pub(crate) data_plane_subnet_filter: String,
  /// AMI type used on default node group when secondary node group (accelerated, Windows, etc) is used
  pub(crate) default_ami_type: ami::AmiType,
  /// Instance types used on default node group when secondary node group (accelerated, Windows, etc) is used
  pub(crate) default_instance_types: Vec<String>,
  pub(crate) enable_cluster_creator_admin_permissions: bool,
  pub(crate) require_efa: bool,
  pub(crate) instance_storage_supported: bool,
  pub(crate) instance_types: Vec<String>,
  pub(crate) reservation: compute::ReservationType,
  pub(crate) reservation_availability_zone: String,
  pub(crate) vpc_name: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Output {
  enable_accelerator: bool,
  enable_nvidia_gpus: bool,
  enable_neuron_devices: bool,
  enable_efa: bool,

  enable_add_ons: bool,
  enable_pod_identity: bool,
  enable_helm: bool,
  enable_public_ecr_helm: bool,

  enable_auto_mode: bool,
  enable_karpenter: bool,
  enable_compute_reservation: bool,
  enable_odcr: bool,
  enable_ml_cbr: bool,

  // Pass through
  add_ons: Vec<add_on::AddOn>,
  ami_type: ami::AmiType,
  cluster_endpoint_public_access: bool,
  cluster_name: String,
  cluster_version: version::ClusterVersion,
  control_plane_subnet_filter: String,
  cpu_arch: compute::CpuArch,
  data_plane_subnet_filter: String,
  default_ami_type: ami::AmiType,
  default_instance_types: Vec<String>,
  enable_cluster_creator_admin_permissions: bool,
  instance_storage_supported: bool,
  instance_types: Vec<String>,
  reservation_availability_zone: String,
  vpc_name: String,
}

impl Default for Inputs {
  fn default() -> Self {
    Inputs {
      accelerator: compute::AcceleratorType::None,
      add_ons: add_on::get_default_add_ons(),
      ami_type: ami::AmiType::AL2023_x86_64_STANDARD,
      cluster_endpoint_public_access: false,
      cluster_name: String::from("example"),
      cluster_version: version::ClusterVersion::K132,
      control_plane_subnet_filter: String::from("*-private-*"),
      compute_scaling: compute::ScalingType::AutoMode,
      cpu_arch: compute::CpuArch::X8664,
      data_plane_subnet_filter: String::from("*-private-*"),
      default_ami_type: ami::AmiType::AL2023_x86_64_STANDARD,
      default_instance_types: vec!["m7a.xlarge".to_string(), "m7i.xlarge".to_string()],
      enable_cluster_creator_admin_permissions: false,
      require_efa: false,
      instance_storage_supported: false,
      instance_types: vec!["m7a.xlarge".to_string(), "m7i.xlarge".to_string()],
      reservation: compute::ReservationType::None,
      reservation_availability_zone: String::from("us-west-2a"),
      vpc_name: String::from("example"),
    }
  }
}

impl Inputs {
  pub fn new() -> Self {
    Self::default()
  }

  pub fn collect(self) -> Result<Output> {
    let inputs = self
      .collect_cluster_settings()?
      .collect_accelerator_type()?
      .collect_require_efa()?
      .collect_reservation_type()?
      .collect_compute_scaling_type()?
      .collect_add_ons()?
      .collect_networking_settings()?
      .collect_cpu_arch()?
      .collect_ami_type()?
      .collect_instance_types()?
      .collect_storage_settings()?
      .collect_default_node_group_settings()?;

    let outputs = inputs.to_output();

    Ok(outputs)
  }

  fn collect_cluster_settings(mut self) -> Result<Inputs> {
    self.cluster_name = Input::with_theme(&ColorfulTheme::default())
      .with_prompt("Cluster name")
      .interact_text()?;

    let cluster_versions = version::ClusterVersion::versions();
    let idx = Select::with_theme(&ColorfulTheme::default())
      .with_prompt("Cluster version")
      .items(&cluster_versions[..])
      .default(0)
      .interact()?;
    self.cluster_version = version::ClusterVersion::from_idx(idx);

    self.cluster_endpoint_public_access = Confirm::with_theme(&ColorfulTheme::default())
      .with_prompt("Enable public access to cluster endpoint")
      .default(false)
      .interact()?;

    self.enable_cluster_creator_admin_permissions = Confirm::with_theme(&ColorfulTheme::default())
      .with_prompt("Enable admin permissions for cluster creator")
      .default(false)
      .interact()?;

    Ok(self)
  }

  fn collect_accelerator_type(mut self) -> Result<Inputs> {
    let all_accelerators = compute::get_accelerator_types();
    let idx = Select::with_theme(&ColorfulTheme::default())
      .with_prompt("Accelerator type")
      .items(&all_accelerators[..])
      .default(0)
      .interact()?;
    self.accelerator = compute::AcceleratorType::from_idx(idx);

    Ok(self)
  }

  fn collect_require_efa(mut self) -> Result<Inputs> {
    match self.accelerator {
      compute::AcceleratorType::Nvidia | compute::AcceleratorType::Neuron => {
        self.require_efa = Confirm::with_theme(&ColorfulTheme::default())
          .with_prompt("Enable EFA support")
          .default(true)
          .interact()?
      }
      _ => {}
    }

    Ok(self)
  }

  fn collect_reservation_type(mut self) -> Result<Inputs> {
    let reservation_types = compute::get_reservation_types(&self.accelerator);
    let idx = Select::with_theme(&ColorfulTheme::default())
      .with_prompt("EC2 capacity reservation")
      .items(&reservation_types[..])
      .default(0)
      .interact()?;
    self.reservation = compute::ReservationType::from_idx(idx);

    Ok(self)
  }

  fn collect_compute_scaling_type(mut self) -> Result<Inputs> {
    let scaling_types = compute::get_compute_scaling_types(&self.reservation);
    let idx = Select::with_theme(&ColorfulTheme::default())
      .with_prompt("Compute autoscaling")
      .items(&scaling_types[..])
      .default(0)
      .interact()?;
    self.compute_scaling = compute::ScalingType::from_idx(idx);

    Ok(self)
  }

  fn collect_add_ons(mut self) -> Result<Inputs> {
    let all_add_ons = add_on::get_add_on_names(&self.compute_scaling);
    let add_ons_idxs = MultiSelect::with_theme(&ColorfulTheme::default())
      .with_prompt("EKS add-on(s)")
      .items(&all_add_ons[..])
      .defaults(&add_on::get_default_add_on_flags())
      .interact()?;

    let add_ons = add_ons_idxs
      .iter()
      .map(|&i| add_on::get_add_on(add_on::AddOnType::from_idx(i)).unwrap())
      .collect::<Vec<add_on::AddOn>>();
    self.add_ons = add_ons;

    Ok(self)
  }

  fn collect_networking_settings(mut self) -> Result<Inputs> {
    self.vpc_name = Input::with_theme(&ColorfulTheme::default())
      .with_prompt("VPC name")
      .with_initial_text("example".to_string())
      .interact_text()?;

    self.control_plane_subnet_filter = Input::with_theme(&ColorfulTheme::default())
      .with_prompt("Control plane subnet filter")
      .with_initial_text("*-private-*".to_string())
      .interact_text()?;

    self.data_plane_subnet_filter = Input::with_theme(&ColorfulTheme::default())
      .with_prompt("Data plane subnet filter")
      .with_initial_text("*-private-*".to_string())
      .interact_text()?;

    if self.reservation != compute::ReservationType::None && self.compute_scaling != compute::ScalingType::Karpenter {
      self.reservation_availability_zone = Input::with_theme(&ColorfulTheme::default())
        .with_prompt("EC2 capacity reservation availability zone")
        .with_initial_text("us-west-2a".to_string())
        .interact_text()?;
    }

    Ok(self)
  }

  fn collect_cpu_arch(mut self) -> Result<Inputs> {
    if !should_collect_arch(&self.compute_scaling, &self.accelerator, self.require_efa) {
      return Ok(self);
    }

    let idx = Select::with_theme(&ColorfulTheme::default())
      .with_prompt("CPU architecture")
      .item("x86-64")
      .item("arm64")
      .default(0)
      .interact()?;
    self.cpu_arch = compute::CpuArch::from_idx(idx);

    Ok(self)
  }

  fn collect_ami_type(mut self) -> Result<Inputs> {
    if self.compute_scaling == compute::ScalingType::AutoMode {
      return Ok(self);
    }

    let ami_types = ami::get_ami_types(&self.accelerator, self.require_efa, &self.cpu_arch);
    let idx = Select::with_theme(&ColorfulTheme::default())
      .with_prompt("AMI type")
      .items(&ami_types[..])
      .default(0)
      .interact()?;
    self.ami_type = ami::AmiType::from_idx(idx);

    Ok(self)
  }

  fn collect_instance_types(mut self) -> Result<Inputs> {
    if self.compute_scaling == compute::ScalingType::AutoMode {
      return Ok(self);
    }

    let instance_type_names =
      compute::get_instance_type_names(&self.cpu_arch, self.require_efa, &self.accelerator, &self.reservation);

    let instance_idxs = MultiSelect::with_theme(&ColorfulTheme::default())
      .with_prompt("Instance type(s)")
      .items(&instance_type_names)
      .interact()?;

    self.instance_types =
      compute::limit_instances_selected(&self.reservation, self.require_efa, instance_type_names, instance_idxs)?;

    Ok(self)
  }

  fn collect_storage_settings(mut self) -> Result<Inputs> {
    self.instance_storage_supported =
      compute::instance_storage_supported(self.instance_types.as_slice(), &self.ami_type);

    Ok(self)
  }

  fn collect_default_node_group_settings(mut self) -> Result<Inputs> {
    self.default_ami_type = ami::get_default_ami_type(&self.ami_type, &self.cpu_arch);

    // Based on the default AMI type selected, set the default instance type(s) for the default node group
    self.default_instance_types = match self.default_ami_type {
      ami::AmiType::AL2023_ARM_64_STANDARD
      | ami::AmiType::BOTTLEROCKET_ARM_64
      | ami::AmiType::BOTTLEROCKET_ARM_64_NVIDIA => {
        vec!["m7g.xlarge".to_string(), "m6g.xlarge".to_string()]
      }
      _ => vec!["m7a.xlarge".to_string(), "m7i.xlarge".to_string()],
    };

    Ok(self)
  }

  pub fn to_output(self) -> Output {
    Output {
      enable_accelerator: self.accelerator != compute::AcceleratorType::None,
      enable_nvidia_gpus: self.accelerator == compute::AcceleratorType::Nvidia,
      enable_neuron_devices: self.accelerator == compute::AcceleratorType::Neuron,
      enable_efa: self.require_efa,

      enable_add_ons: !self.add_ons.is_empty(),
      enable_pod_identity: self
        .add_ons
        .iter()
        .any(|a| a.configuration.is_some() && a.configuration.as_ref().unwrap().pod_identity_role_arn.is_some()),
      enable_helm: should_enable_helm(&self.accelerator, &self.compute_scaling, self.require_efa),
      enable_public_ecr_helm: should_enable_public_ecr_helm(&self.accelerator, &self.compute_scaling),

      enable_auto_mode: self.compute_scaling == compute::ScalingType::AutoMode,
      enable_karpenter: self.compute_scaling == compute::ScalingType::Karpenter,
      enable_compute_reservation: self.reservation != compute::ReservationType::None,
      enable_odcr: self.reservation == compute::ReservationType::OnDemandCapacityReservation,
      enable_ml_cbr: self.reservation == compute::ReservationType::MlCapacityBlockReservation,

      // Pass through
      add_ons: self.add_ons,
      ami_type: self.ami_type,
      cluster_endpoint_public_access: self.cluster_endpoint_public_access,
      cluster_name: self.cluster_name,
      cluster_version: self.cluster_version,
      control_plane_subnet_filter: self.control_plane_subnet_filter,
      cpu_arch: self.cpu_arch,
      data_plane_subnet_filter: self.data_plane_subnet_filter,
      default_ami_type: self.default_ami_type,
      default_instance_types: self.default_instance_types,
      enable_cluster_creator_admin_permissions: self.enable_cluster_creator_admin_permissions,
      instance_storage_supported: self.instance_storage_supported,
      instance_types: self.instance_types,
      reservation_availability_zone: self.reservation_availability_zone,
      vpc_name: self.vpc_name,
    }
  }
}

fn should_collect_arch(
  scaling_type: &compute::ScalingType,
  accelerator: &compute::AcceleratorType,
  require_efa: bool,
) -> bool {
  // Set on Auto Mode/Karpenter NodeClass
  if *scaling_type == compute::ScalingType::Karpenter || *scaling_type == compute::ScalingType::AutoMode {
    return false;
  }

  // Inf/Trn instances only support x86-64 at this time
  if *accelerator == compute::AcceleratorType::Neuron {
    return false;
  }

  if *accelerator == compute::AcceleratorType::Nvidia && require_efa {
    return false;
  }

  true
}

// Standard Helm resources are:
// - NVIDIA device plugin
// - EFA device plugin
fn should_enable_helm(
  accelerator: &compute::AcceleratorType,
  compute: &compute::ScalingType,
  require_efa: bool,
) -> bool {
  // Auto Mode bundles the NVIDIA device plugin and EFA device plugin
  if compute == &compute::ScalingType::AutoMode {
    return false;
  }
  if accelerator == &compute::AcceleratorType::Nvidia || require_efa {
    return true;
  }
  false
}

// Public ECR Helm resources are:
// - Neuron Helm chart
// - Karpenter Helm chart
fn should_enable_public_ecr_helm(accelerator: &compute::AcceleratorType, compute: &compute::ScalingType) -> bool {
  // Auto Mode bundles the Neuron Helm chart and Karpenter Helm chart (equivalents)
  if compute == &compute::ScalingType::AutoMode {
    return false;
  }
  if accelerator == &compute::AcceleratorType::Neuron || compute == &compute::ScalingType::Karpenter {
    return true;
  }
  false
}
