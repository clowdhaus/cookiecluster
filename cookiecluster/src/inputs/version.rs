use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter, EnumString, IntoStaticStr};

#[derive(Debug, EnumIter, Display, EnumString, IntoStaticStr, PartialEq, Serialize, Deserialize)]
pub enum ClusterVersion {
  #[strum(serialize = "1.32")]
  #[serde(rename = "1.32")]
  K132,
  #[strum(serialize = "1.31")]
  #[serde(rename = "1.31")]
  K131,
  #[strum(serialize = "1.30")]
  #[serde(rename = "1.30")]
  K130,
  #[strum(serialize = "1.29")]
  #[serde(rename = "1.29")]
  K129,
  #[strum(serialize = "1.28")]
  #[serde(rename = "1.28")]
  K128,
  #[strum(serialize = "1.27")]
  #[serde(rename = "1.27")]
  K127,
}

impl ClusterVersion {
  #[cfg(not(tarpaulin_include))]
  pub fn versions() -> Vec<ClusterVersion> {
    ClusterVersion::iter().collect()
  }

  #[cfg(not(tarpaulin_include))]
  pub fn from_idx(idx: usize) -> ClusterVersion {
    ClusterVersion::iter().nth(idx).unwrap()
  }
}
