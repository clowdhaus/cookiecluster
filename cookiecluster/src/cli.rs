use std::{fs, path::Path};

use anstyle::{AnsiColor, Color, Style};
use anyhow::Result;
use clap::{Parser, builder::Styles};
use clap_verbosity_flag::{InfoLevel, Verbosity};
use handlebars::Handlebars;
use serde_json::value::Map;

use crate::inputs::Output;

/// Styles for CLI
fn get_styles() -> Styles {
  Styles::styled()
    .header(
      Style::new()
        .bold()
        .underline()
        .fg_color(Some(Color::Ansi(AnsiColor::Blue))),
    )
    .literal(Style::new().bold().fg_color(Some(Color::Ansi(AnsiColor::Cyan))))
    .usage(
      Style::new()
        .bold()
        .underline()
        .fg_color(Some(Color::Ansi(AnsiColor::Blue))),
    )
    .placeholder(Style::new().bold().fg_color(Some(Color::Ansi(AnsiColor::Magenta))))
}

/// Cookiecluster - A CLI to generate EKS cluster definitions in Terraform.
///
/// Based on a few inputs, cookiecluster will guide you through the process of designing a cluster that fits your
/// requirements using the appropriate values. It does not require any AWS credentials; it is merely a glorified
/// templating engine for EKS clusters defined in Terraform.
#[derive(Debug, Parser)]
#[command(author, about, version)]
#[command(propagate_version = true)]
#[command(styles=get_styles())]
pub struct Cli {
  #[clap(flatten)]
  pub verbose: Verbosity<InfoLevel>,
}

impl Cli {
  pub fn write(self, output: &Output) -> Result<()> {
    let handlebars = crate::register_handlebars()?;

    fs::write(Path::new("eks.tf"), render_value("eks", output, &handlebars)?)?;
    fs::write(Path::new("main.tf"), render_value("main", output, &handlebars)?)?;

    let helm = render_value("helm", output, &handlebars)?;
    if !helm.is_empty() {
      fs::write(Path::new("helm.tf"), helm)?;
    }

    let vars = render_value("variables", output, &handlebars)?;
    if !vars.is_empty() {
      fs::write(Path::new("variables.tf"), vars)?;
    }

    match std::process::Command::new("terraform").arg("fmt").arg(".").output() {
      Ok(_) => tracing::trace!("Terraform files have been formatted"),
      _ => tracing::trace!("Terraform executable not found. Skipping formatting."),
    };

    Ok(())
  }
}

fn render_value(name: &str, output: &Output, handlebars: &Handlebars) -> Result<String> {
  let mut data = Map::new();
  data.insert("inputs".to_string(), handlebars::to_json(output));

  handlebars.render(name, &data).map_err(Into::into)
}

#[cfg(test)]
mod tests {

  use super::*;
  use crate::inputs::{Inputs, add_on, ami, compute, version};

  fn render(output: Output, dir_name: &str) -> Result<()> {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path(format!("./snapshots/{dir_name}"));
    let _guard = settings.bind_to_scope();

    for tpl in ["eks", "main", "helm", "variables"] {
      let handlebars = crate::register_handlebars()?;
      let rendered = render_value(tpl, &output, &handlebars)?;
      insta::assert_snapshot!(tpl, rendered);
    }

    Ok(())
  }

  #[test]
  fn snapshot_default() {
    // Defaults to AL2023
    let inputs = Inputs::default();

    render(inputs.to_output(), "default").unwrap();
  }

  #[test]
  fn snapshot_al2023_x86_64() {
    let inputs = Inputs {
      ami_type: ami::AmiType::AL2023_x86_64_STANDARD,
      compute_scaling: compute::ScalingType::None,
      ..Inputs::default()
    };

    render(inputs.to_output(), "al2023-x86-64").unwrap();
  }

  #[test]
  fn snapshot_al2023_arm64() {
    let inputs = Inputs {
      ami_type: ami::AmiType::AL2023_ARM_64_STANDARD,
      compute_scaling: compute::ScalingType::None,
      instance_types: vec!["m7g.xlarge".to_string(), "m6g.xlarge".to_string()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "al2032-arm64").unwrap();
  }

  #[test]
  fn snapshot_bottlerocket_x8664() {
    let inputs = Inputs {
      ami_type: ami::AmiType::BOTTLEROCKET_x86_64,
      compute_scaling: compute::ScalingType::None,
      ..Inputs::default()
    };

    render(inputs.to_output(), "bottlerocket-x86-64").unwrap();
  }

  #[test]
  fn snapshot_bottlerocket_arm64() {
    let inputs = Inputs {
      ami_type: ami::AmiType::BOTTLEROCKET_ARM_64,
      compute_scaling: compute::ScalingType::None,
      instance_types: vec!["m7g.xlarge".to_string(), "m6g.xlarge".to_string()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "bottlerocket-arm64").unwrap();
  }

  #[test]
  fn snapshot_nvidia() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Nvidia,
      ami_type: ami::AmiType::AL2023_x86_64_NVIDIA,
      compute_scaling: compute::ScalingType::None,
      instance_types: vec!["g5.4xlarge".to_owned()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "nvidia").unwrap();
  }

  #[test]
  fn snapshot_nvidia_efa() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Nvidia,
      ami_type: ami::AmiType::AL2023_x86_64_NVIDIA,
      compute_scaling: compute::ScalingType::None,
      require_efa: true,
      instance_storage_supported: true,
      instance_types: vec!["p5.48xlarge".to_owned()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "nvidia_efa").unwrap();
  }

  #[test]
  fn snapshot_nvidia_efa_odcr() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Nvidia,
      ami_type: ami::AmiType::AL2023_x86_64_NVIDIA,
      compute_scaling: compute::ScalingType::None,
      require_efa: true,
      instance_storage_supported: true,
      instance_types: vec!["p5.48xlarge".to_owned()],
      reservation: compute::ReservationType::OnDemandCapacityReservation,
      ..Inputs::default()
    };

    render(inputs.to_output(), "nvidia-efa-odcr").unwrap();
  }

  #[test]
  fn snapshot_nvidia_efa_cbr() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Nvidia,
      ami_type: ami::AmiType::AL2023_x86_64_NVIDIA,
      compute_scaling: compute::ScalingType::None,
      require_efa: true,
      instance_storage_supported: true,
      instance_types: vec!["p5.48xlarge".to_owned()],
      reservation: compute::ReservationType::MlCapacityBlockReservation,
      ..Inputs::default()
    };

    render(inputs.to_output(), "nvidia-efa-cbr").unwrap();
  }

  #[test]
  fn snapshot_nvidia_cbr() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Nvidia,
      ami_type: ami::AmiType::AL2023_x86_64_NVIDIA,
      compute_scaling: compute::ScalingType::None,
      require_efa: false,
      instance_storage_supported: true,
      instance_types: vec!["p5.48xlarge".to_owned()],
      reservation: compute::ReservationType::MlCapacityBlockReservation,
      ..Inputs::default()
    };

    render(inputs.to_output(), "nvidia-cbr").unwrap();
  }

  #[test]
  fn snapshot_neuron() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Neuron,
      ami_type: ami::AmiType::AL2023_x86_64_NEURON,
      compute_scaling: compute::ScalingType::None,
      instance_types: vec!["inf2.xlarge".to_owned()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "neuron").unwrap();
  }

  #[test]
  fn snapshot_neuron_efa() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Neuron,
      ami_type: ami::AmiType::AL2023_x86_64_NEURON,
      compute_scaling: compute::ScalingType::None,
      require_efa: true,
      instance_storage_supported: true,
      instance_types: vec!["trn1n.32xlarge".to_owned()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "neuron-efa").unwrap();
  }

  #[test]
  fn snapshot_neuron_cbr() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Neuron,
      ami_type: ami::AmiType::AL2023_x86_64_NEURON,
      compute_scaling: compute::ScalingType::None,
      require_efa: false,
      instance_storage_supported: true,
      instance_types: vec!["trn1.32xlarge".to_owned()],
      reservation: compute::ReservationType::MlCapacityBlockReservation,
      ..Inputs::default()
    };

    render(inputs.to_output(), "neuron-cbr").unwrap();
  }

  #[test]
  fn snapshot_al2023_instance_storage() {
    let inputs = Inputs {
      ami_type: ami::AmiType::AL2023_ARM_64_STANDARD,
      compute_scaling: compute::ScalingType::None,
      instance_storage_supported: true,
      instance_types: vec!["m7gd.2xlarge".to_owned()],
      ..Inputs::default()
    };

    render(inputs.to_output(), "al2023-instance-storage").unwrap();
  }

  #[test]
  fn snapshot_karpenter() {
    let inputs = Inputs {
      compute_scaling: compute::ScalingType::Karpenter,
      ..Inputs::default()
    };

    render(inputs.to_output(), "karpenter").unwrap();
  }

  #[test]
  fn snapshot_karpenter_odcr() {
    let inputs = Inputs {
      accelerator: compute::AcceleratorType::Nvidia,
      ami_type: ami::AmiType::AL2023_x86_64_NVIDIA,
      compute_scaling: compute::ScalingType::Karpenter,
      instance_storage_supported: true,
      instance_types: vec!["p5e.48xlarge".to_owned()],
      reservation: compute::ReservationType::OnDemandCapacityReservation,
      ..Inputs::default()
    };

    render(inputs.to_output(), "karpenter-odcr").unwrap();
  }

  #[test]
  fn snapshot_auto_mode() {
    let inputs = Inputs {
      compute_scaling: compute::ScalingType::AutoMode,
      add_ons: vec![],
      ..Inputs::default()
    };

    render(inputs.to_output(), "auto-mode").unwrap();
  }

  #[test]
  fn snapshot_enable_all() {
    let inputs = Inputs {
      cluster_name: "cookiecluster".to_string(),
      cluster_version: version::ClusterVersion::K128,
      cluster_endpoint_public_access: true,
      compute_scaling: compute::ScalingType::None,
      control_plane_subnet_filter: "*-intra-*".to_string(),
      data_plane_subnet_filter: "*-data-*".to_string(),
      enable_cluster_creator_admin_permissions: true,
      vpc_name: "cookiecluster".to_string(),
      ..Inputs::default()
    };

    render(inputs.to_output(), "enable-all").unwrap();
  }

  #[test]
  fn snapshot_all_add_ons() {
    let inputs = Inputs {
      add_ons: add_on::_get_all_add_ons(),
      compute_scaling: compute::ScalingType::None,
      ..Inputs::default()
    };

    render(inputs.to_output(), "all-add-ons").unwrap();
  }
}
