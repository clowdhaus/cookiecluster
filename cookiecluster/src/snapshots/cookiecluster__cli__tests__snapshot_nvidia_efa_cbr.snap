---
source: cookiecluster/src/cli.rs
expression: rendered
---
/*
## NVIDIA K8s Device Plugin

The NVIDIA K8s device plugin, https://github.com/NVIDIA/k8s-device-plugin, will need to
be installed in the cluster in order to mount and utilize the GPUs in your pods. Add the
following affinity rule to your device plugin Helm chart values to ensure the device
plugin runs on nodes that have GPUs present (as identified via the MNG
labels provided below):

```yaml
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
        - matchExpressions:
          - key: 'nvidia.com/gpu.present'
            operator: In
            values:
              - 'true'
```

By default, the NVIDIA K8s device values already contian a toleration that matches the taint applied
to the node group below.

## EFA K8s Device Plugin

The EFA K8s device plugin, https://github.com/aws/eks-charts/tree/master/stable/aws-efa-k8s-device-plugin,
will need to  be installed in the cluster in order to mount and utilize the EFA devcies
in your pods. Add the following node selector and toleration to your device plugin Helm chart
values to ensure the device plugin runs on nodes that have EFA devices present
(as identified via the MNG label and taint provided below):

```yaml
nodeSelector:
  vpc.amazonaws.com/efa.present: 'true'
tolerations:
  - key: nvidia.com/gpu
    operator: Exists
    effect: NoSchedule
```
*/
################################################################################
# EKS Cluster
################################################################################

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 20.24"

  cluster_name    = "example"
  cluster_version = "1.30"

  cluster_addons = {
    coredns = {}
    eks-pod-identity-agent = {}
    kube-proxy = {}
    vpc-cni = {}
  }

  # Add security group rules on the node group security group to
  # allow EFA traffic
  enable_efa_support = true

  vpc_id                   = data.aws_vpc.this.id
  control_plane_subnet_ids = data.aws_subnets.control_plane.ids
  subnet_ids               = data.aws_subnets.data_plane.ids

  eks_managed_node_groups = {
    # This node group is for core addons such as CoreDNS
    default = {
      ami_type = "AL2023_x86_64_STANDARD"
      instance_types = [
        "m7a.xlarge",
        "m7i.xlarge",
      ]

      min_size     = 2
      max_size     = 3
      desired_size = 2
    }
    gpu = {
      ami_type       = "AL2_x86_64_GPU"
      instance_types = ["p5.48xlarge" ]

      min_size     = 2
      max_size     = 5
      desired_size = 2

      pre_bootstrap_user_data = <<-EOT
        #!/usr/bin/env bash

        # Mount instance store volumes in RAID-0 for Kubelet and Containerd (raid0)
        # https://github.com/awslabs/amazon-eks-ami/blob/master/doc/USER_GUIDE.md#raid-0-for-kubelet-and-containerd-raid0
        /bin/setup-local-disks raid0
      EOT

      # Add security group rules on the node group security group to
      # allow EFA traffic
      enable_efa_support = true

      labels = {
        "vpc.amazonaws.com/efa.present" = "true"
        "nvidia.com/gpu.present"        = "true"
      }

      taints = {
        # Ensure only GPU workloads are scheduled on this node group
        gpu = {
          key    = "nvidia.com/gpu"
          value  = "true"
          effect = "NO_SCHEDULE"
        }
      }

      # Capacity reservations are restricted to a single availability zone
      subnet_ids = data.aws_subnets.data_plane_reservation.ids
      
      # ML capacity block reservation
      capacity_type = "CAPACITY_BLOCK"
      instance_market_options = {
        market_type = "capacity-block"
      }
      capacity_reservation_specification = {
        capacity_reservation_target = {
          capacity_reservation_id = var.capacity_reservation_id
        }
      }
    }
  }

  tags = module.tags.tags
}

################################################################################
# Tags - Replace with your own tags implementation
################################################################################

module "tags" {
  source  = "clowdhaus/tags/aws"
  version = "~> 1.1"

  application = "cookiecluster"
  environment = "nonprod"
  repository  = "github.com/clowdhaus/cookiecluster"
}

################################################################################
# Variables - Required input
################################################################################

variable "capacity_reservation_id" {
  description = "The ID of the ML capacity block reservation in which to run the instance(s)"
  type        = string
}

################################################################################
# VPC & Subnet data sources
################################################################################

data "aws_vpc" "this" {
  filter {
    name   = "tag:Name"
    values = ["example"]
  }
}

data "aws_subnets" "control_plane" {
  filter {
    name   = "tag:Name"
    values = ["*-private-*"]
  }
}

data "aws_subnets" "data_plane" {
  filter {
    name   = "tag:Name"
    values = ["*-private-*"]
  }
}

data "aws_subnets" "data_plane_reservation" {
  filter {
    name   = "tag:Name"
    values = ["*-private-*"]
  }

  # Capacity reservations are restricted to a single availability zone
  filter {
    name = "availability-zone"
    values = ["us-west-2a"]
  }
}
